src/Authenticator/ChangePasswordHandler.php

```diff
@@ -0,0 +1,149 @@
+                    'endpoints' => [
+                        'login' => $this->Link('mfa/login/{urlSegment}'),
+                        'complete' => $this->Link(),
+                    ],
+                    'shouldRedirect' => false,
+                ])
+            );
+        } catch (Throwable $exception) {
+            // If we don't have a valid member we shouldn't be here...
+            return $this->redirectBack();
+        }
+    }
+
+    public function mfa()
+    {
+        $store = $this->getStore();
+        if (!$store || !$store->getMember()) {
+            return $this->redirectBack();
+        }
+
+        $this->applyRequirements();
+
+        return [
+            'Form' => $this->renderWith($this->getViewerTemplates()),
+            'ClassName' => 'mfa',
+        ];
+    }
+
+    public function startMFACheck(HTTPRequest $request)
+    {
+        $store = $this->getStore();
+        $member = $store->getMember();
+
+        // If we don't have a valid member we shouldn't be here...
+        if (!$member) {
+            return $this->jsonResponse(['message' => 'Forbidden'], 403);
+        }
+
+        // Use the provided trait method for handling login
+        $response = $this->createStartLoginResponse(
+            $store,
+            Injector::inst()->get(MethodRegistry::class)->getMethodByURLSegment($request->param('Method'))
+        );
+
+        // Ensure detail is saved to the store
+        $store->save($request);
+
+        // Respond with our method
+        return $response;
+    }
+
+    public function verifyMFACheck(HTTPRequest $request)
+    {
+        $store = $this->getStore();
+
+        try {
+            $result = $this->verifyLoginRequest($store, $request);
+        } catch (InvalidMethodException $e) {
+            // Invalid method usually means a timeout. A user might be trying to verify before "starting"
+            return $this->jsonResponse(['message' => 'Forbidden'], 403);
+        }
+
+        if (!$result->isSuccessful()) {
+            return $this->jsonResponse([
+                'message' => $result->getMessage(),
+            ], 401);
+        }
+
+        if (!$this->isLoginComplete($store)) {
+            return $this->jsonResponse([
+                'message' => 'Additional authentication required',
+            ], 202);
+        }
+
+        $this->getRequest()->getSession()->set('MultiFactorAuthenticated', true);
+        $store->clear($request);
+
+        return $this->jsonResponse([
+            'message' => 'Multi factor authenticated',
+        ], 200);
+    }
+
+    public function changepassword()
+    {
+        $session = $this->getRequest()->getSession();
+        $hash = $session->get('AutoLoginHash');
+        $member = Member::member_from_autologinhash($hash);
+        if ((
+            $hash
+            && $member
+            && $member->RegisteredMFAMethods()->exists()
+            && !$session->get('MultiFactorAuthenticated')
+        )) {
```

Comment:
You can remove one of these brackets. You may also want to use `$member->getDefaultRegisteredMethod()` in the condition here over the exists check. A member will have a default registered method from their initial setup.

Reply:
It is a bit of a micro-optimisation, but I feel that exists will be faster as it does not inflate an object.

Reply:
OK